Fix TSan bug with C calls that take many arguments

OlivierNicole opened the pull request:
Currently a draft as it is only implemented for x86_64, and depends on #14213.

This addresses a wrong signaling to TSan in `caml_c_call_stack_args`, both regarding the routine's stack frame itself, but also the stack frame constituted by the additional arguments on the C stack (see individual commit messages).

gasche commented:
(This PR could be rebased now that we merged the unwinding fixes.)

Here is what I understand about `caml_c_call_stack_args`, the focus of the current PR:

- It is designed for functions that take more arguments than the native C convention allows through registers, the extra arguments are to be pushed on the C call stack
- It is implemented by two nested functions, first `caml_c_call_stack_args` which follows the OCaml calling convention and switches internally to the C calling convention, with which it calls `caml_c_call_copy_stack_args` which pushes argument to the stack, then call the intended C function. For book-keeping purposes, `..._copy_stack_args` stores the current stack pointer before pushing the arguments (and restores it when the user-provided C function returns).

Here is what I understand about the changes you propose:

- In `caml_c_call_stack_args` you use standard TSan instrumentation, similar in spirit to how `caml_c_call` does it. This is reasonably simple and looks correct.
- You also add TSan instrumentation in `caml_c_call_copy_stack_args` , but there I am less sure of what is going on.

gasche commented on runtime/amd64.S:
```diff
@@ -764,6 +777,18 @@ CFI_STARTPROC
+       the usual TSAN_* macros which assume we start on the OCaml stack. */
+        movq    (FRAME_POINTER_SIZE)(%rsp), C_ARG_1
+        C_call  (GCALL(caml_tsan_func_entry_asm));
+        TSAN_RESTORE_CALLER_REGS
```
From a distance this looks like the code we would write when we are on the OCaml stack, following OCaml calling conventions. Here my understanding is that we are already on the C stack, following the C calling convention (and our caller context is known and we can tell which caller-saved registers it actually relies on). Are we sure we need the CALLER_REGS stuff?

gasche replied:
Note: when I ask about they are necessary, my idea is not to suggest how to make the code faster (this is not the focus here), but rather to make it simpler and easier to understand. Currently all the calls to `*_CALLER_REGS` come when we are on the OCaml stack, and this regularity is soothing.

OlivierNicole replied:
Ah, yes, I got confused. I need to preserve the arguments passed to `caml_c_call_copy_stack_args` but not like this.

gasche commented on runtime/amd64.S:
```diff
@@ -783,6 +808,16 @@ LBL(106):
         CFI_DEF_CFA_REGISTER(DW_REG_rsp)
         popq    %rbp; CFI_ADJUST(-8)
+#ifdef WITH_THREAD_SANITIZER
+    /* Save non-callee-saved registers %rax and %xmm0 before C call */
```
Same questions here on return: are we sure we need this extra book-keeping to save various registers? Who needs them?

OlivierNicole replied:
My intention here is to preserve the return value of the C call in case the calling code uses it—similarly to what `caml_c_call` does.

gasche replied:
So this block of code would be used whenever we call a C function that must not clobber arbitrary return values. Could we maybe consider a `TSAN_{SAVE,RESTORE}_RETURN_REGS` macro for this, to be used by `caml_c_call`, `caml_c_call_stack_args`, `caml_c_call_copy_stack_args` and possibly also `caml_start_program`?

OlivierNicole replied:
Makes sense to me

gasche commented on runtime/amd64.S:
```diff
@@ -779,15 +785,37 @@ CFI_STARTPROC
+        movq          4*8(%r15),%rdx
+        movq          5*8(%r15),%rcx
+        movq          6*8(%r15),%r8
+        movq          7*8(%r15),%r9
```
There is this new stuff in `caml_c_call_copy_stack_args`, I don't understand where it comes from. Can you explain?

OlivierNicole replied:
This is using the GC register saving buckets to save the routine arguments before calling into C. It is similar to `TSAN_SAVE_CALLER_REGS`, without the unnecessary instructions. It does not make assumptions about the stack.

gasche replied:
I don't like the fact that the complex logic of using `gc_buckets` is now duplicated in several places, and furthermore that one instance is within already-tricky code.

My non-expert intuition is that it should be possible to do whatever register-saving is necessary from `call_c_stack_args` already, and not have to deal with this stuff from within the callee function `call_c_copy_stack_args`. But even if this intuition is wrong, I would prefer to see some dedicated macros for that complex boilerplate stuff, possibly shared with the definitions of `TSAN_{SAVE,RESTORE}_CALLER_REGS`, but in any case somewhere *away* from this function body.

OlivierNicole replied:
I understand. I now think that using `TSAN_{SAVE,RESTORE}_CALLER_REGS` in this function was a good compromise. It was only missing instructions to save r12 and r13.

OlivierNicole replied:
Self-correction: actually r12 and r13 are callee-saved.

OlivierNicole replied:
P.S. 

> My non-expert intuition is that it should be possible to do whatever register-saving is necessary from `call_c_stack_args` already, and not have to deal with this stuff from within the callee function `call_c_copy_stack_args`.

I don’t see a way to do it simply, unfortunately. But maybe I’m missing something.

gasche replied:
Thinking about this more, I think I was maybe thinking that `..._copy_stack_args` would take all its arguments through the stack, so should not need much saving, but in fact it takes register arguments as well, which explains why caller-saving is required no matter where we come from.

gasche commented:
I have asked enough questions by now and I think that the code of the PR is converging to something reasonably simple and well-explained. I see that you added support for other architectures thanks to Miod. Would you be willing to factorize the {PUSH,POP}_RETURN_REGS logic for these other architectures as well? (Or maybe there is a reason why domain experts prefer to repeat it?)

OlivierNicole commented:
Yes, credit goes to Miod Vallat for the other architectures.

I factorized register saving for ARM. As Miod pointed out to me in a private discussion, `caml_c_call_copy_stack_args` doesn’t exist on other architectures, so there is nothing to do there. Do you think we should still introduce similar macros?

I tried to reuse a CI job created by @xavierleroy (for [this PR](https://github.com/ocaml/ocaml/pull/13594#issuecomment-2886837645)) to test TSan on linux ARM64, macOS ARM64, and POWER. However, the jobs don’t seem to work. I’ll need some time to look into it.

gasche commented:
I would be in favor of introducing the macros nonetheless, because (1) they can be used later on, (2) they make the code of `caml_c_call` simpler to read, (3) there is the opportunity to reuse them for `caml_start_program` (but this is not strictly necessary as it knows more about its return values), and (4) keeping the runtime code in sync helps port changes across architectures.

OlivierNicole committed:
Update changelog

gasche commented on runtime/arm64.S:
```diff
@@ -633,6 +658,11 @@ FUNCTION(caml_c_call_stack_args)
+        TSAN_PUSH_RETURN_REGS /* Save returned value. */
+        TSAN_EXIT_FUNCTION
+        TSAN_POP_RETURN_REGS
+#endif
```
When you have a TSAN_-only sequence, the `ifdef` is technically not necessary because these macros are defined to be empty when TSan is disabled. The style that use use in the rest of the file is to avoid the conditionals in that case.

OlivierNicole replied:
Thank you, I had missed this one.

gasche reviewed (approved):
I have checked that this should only affect TSan users (there are no runtime-assembly change when TSan is disabled), and I understand the reason for the changes and believe that they leave the code in a better state than they found it.

I am unfamiliar with most of these assemblies, but I trust the expertise of the co-authors to do something reasonable.

I left a minor/nitpicky comment, but @OlivierNicole I think you should feel free to merge after you have decided whether to take it into account or not -- assuming that you are satisfied with the amount of testing that happened.

OlivierNicole commented:
Thank you for the review. I’ll merge when/if the [CI job](https://ci.inria.fr/ocaml/job/precheck-tsan/8/) shows no errors (well, no more errors than [trunk](https://ci.inria.fr/ocaml/job/precheck-tsan/7/)).

OlivierNicole commented:
There turned out to be problems in register saving for ARM and s390x, for which Miod kindly provided a fix again. Let’s see what the [new testing job](https://ci.inria.fr/ocaml/job/precheck-tsan/9/) says…

OlivierNicole commented:
It looks like a recent change on trunk has incremented a thread ID by 1 in all TSan tests.

gadmm commented:
For better error handling in `caml_domain_spawn`, the backup thread for domain 0 is now spawned before the main thread of domain 1. Could this affect the thread ID you are seeing?

OlivierNicole commented:
IIRC previously the backup thread was only spawned upon the first `Domain.spawn`? In that case yes, it explains it.

OlivierNicole commented:
I intend to do this minor fix on the tests as soon as possible; there is also a tiny discrepancy remianing on POWER in the Jenkins CI.

gadmm commented:
> IIRC previously the backup thread was only spawned upon the first `Domain.spawn`? In that case yes, it explains it.

This is still the case, however the order in which this is done has been switched.

OlivierNicole commented:
Ah, ok, that works too.

OlivierNicole committed:
Update TSan tests after #12410

OlivierNicole committed:
Add missing TSan accounting in caml_c_call_stack_args for x86_64

The absence of these calls to TSan resulted in its shadow stack being
offset and deriving over time, causing wrong backtraces and crashes (due
to the complete absence of bound checks in TSan).

Commit:
Fix caml_c_call_stack_args for ARM64, POWER, RISC-V and s390x

OlivierNicole commented:
I took the liberty to split off this PR. The POWER parts that are proving thorny I’ve put in #14302. The rest, that we believe to be good, remains here. My understanding is that we can merge this.

OlivierNicole commented:
The CI shows nothing unexpected. Merging.

OlivierNicole commented:
@Octachron Possible candidate for 5.4.1 (The runtime is not affected by this PR when TSan is disabled.)

Octachron commented:
@OlivierNicole , I have the impression that we also want #14082 as part of the dependency cone of this PR. Do you agree?

OlivierNicole commented:
Oh! Yes, definitely.
